src/io/flutter/inspector/EvalOnDartLibrary.java

```diff
@@ -0,0 +1,185 @@
+import org.dartlang.vm.service.consumer.EvaluateConsumer;
+import org.dartlang.vm.service.consumer.GetIsolateConsumer;
+import org.dartlang.vm.service.consumer.GetObjectConsumer;
+import org.dartlang.vm.service.element.*;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Collection;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * Invoke methods from a specified Dart library using the observatory protocol.
+ */
+public class EvalOnDartLibrary implements Disposable {
+  private final IsolatesInfo.IsolateInfo isolateInfo;
+  private final VmService vmService;
+  private final String libraryName;
+  final CompletableFuture<LibraryRef> libraryRef;
+  private final Alarm myRequestsScheduler;
+
+  private static final Logger LOG = Logger.getInstance(EvalOnDartLibrary.class);
+
+  public EvalOnDartLibrary(String libraryName, FlutterDebugProcess debugProcess, VmService vmService) {
+    this.vmService = vmService;
+    final Collection<IsolatesInfo.IsolateInfo> isolates = debugProcess.getIsolateInfos();
+    assert (isolates.size() == 1);
+    isolateInfo = Iterables.get(isolates, 0);
+    this.libraryName = libraryName;
+    this.myRequestsScheduler = new Alarm(Alarm.ThreadToUse.POOLED_THREAD, this);
+    libraryRef = new CompletableFuture<>();
+    initialize();
+  }
+
+  public String getIsolateId() {
+    return isolateInfo.getIsolateId();
+  }
+
+  public void dispose() {
+    myRequestsScheduler.dispose();
+    // TODO(jacobr): complete all pending futures as cancelled?
+  }
+
+  /**
+   * Workaround until the version of the VmService 'eval' method supports
+   * the scope argument.
+   */
+  private void callVmServiceRequest(VmService vmService, String methodName, JsonObject params, EvaluateConsumer consumer) {
+    try {
+      Method method = ReflectionUtil.getDeclaredMethod(Class.forName("org.dartlang.vm.service.VmServiceBase"), "request", String.class, JsonObject.class, Consumer.class);
+      if (method == null) {
+        throw new RuntimeException("Cannot find method 'request'");
+      }
+      method.setAccessible(true);
+      method.invoke(vmService, methodName, params, consumer);
+    }
+    catch (IllegalAccessException | InvocationTargetException | ClassNotFoundException e) {
+      throw new RuntimeException((e.toString()));
+    }
+  }
+
+  public CompletableFuture<InstanceRef> eval(String expression, Map<String, String> scope) {
+    final CompletableFuture<InstanceRef> future = new CompletableFuture<>();
+    myRequestsScheduler.addRequest(() -> {
+      libraryRef.thenAcceptAsync((LibraryRef ref) -> {
+        evaluateHelper(isolateInfo.getIsolateId(), ref.getId(), expression, scope,
+                       new EvaluateConsumer() {
+                         @Override
+                         public void onError(RPCError error) {
+                           // TODO(jacobr): what is the right way to log this case?
```

Comment:
Are you wondering if an error here is serious enough to bring to the user's attention? If so, then maybe an error balloon in the style of IntelliJ would work.

Reply:
Removed the TODO as the LOG.error call is sufficient for this unlikely exceptional case.